File: test093-msgpack/py/write_out.py
import io
import sys

import msgpack
import numpy as np

def main():
    if len(sys.argv) == 2:
        dump_to_file(sys.argv[1])
    else:
        dump_to_stdout()

def dump_to_stdout():
    dump(sys.stdout.detach())

def dump_to_file(path):
    with open(path, 'wb') as dest:
        dump(dest)

def dump(stream):
    num_frames = 50
    num_points = 60000
    dimension = 3
    sigma = 0.01

    metadata = {
        'name': 'Lorem ipsum dolor sit amet',
        'num_points': num_points,
        'dimension': dimension
    }

    packer = msgpack.Packer(use_single_float=True)
    stream.write(packer.pack(metadata))
    stream.write(packer.pack_array_header(num_frames))
    for points in generate_frames(num_frames, num_points, dimension, sigma):
        points_data = points.astype(np.float32).tobytes()
        stream.write(packer.pack(points_data))

def generate_frames(num_frames, num_points, dimension, sigma):
    points = np.random.uniform(-1, 1, size=(num_points, dimension))
    for _ in range(num_frames):
        points += np.random.normal(0, sigma, size=points.shape)
        yield points

if __name__ == '__main__':
    main()

File: test093-msgpack/write_out.cc
#include <cstddef>
#include <cstdint>
#include <fstream>
#include <ostream>
#include <random>
#include <string>
#include <vector>

#include <dak.hpp>
#include <msgpack.hpp>
#include <range.hpp>

namespace units
{
    using point = dak::point<double, dak::dim::length, 3>;
    using displacement = dak::vector<double, dak::dim::length, 3>;
}

class random_walk_simulator
{
    double const sigma = 0.01;

  public:
    explicit random_walk_simulator(std::size_t num_points, std::uint32_t seed)
        : points_(num_points), engine_{seed}
    {
        initialize_points();
    }

    range::range<units::point const*> points() const
    {
        return {points_.data(), points_.data() + points_.size()};
    }

    void step_forward()
    {
        std::normal_distribution<double> move{0, sigma};
        for (units::point& point : points_) {
            point += units::displacement{move(engine_), move(engine_), move(engine_)};
        }
    }

  private:
    void initialize_points()
    {
        std::uniform_real_distribution<double> coord{-1, 1};
        for (units::point& point : points_) {
            point = units::point{coord(engine_), coord(engine_), coord(engine_)};
        }
    }

  private:
    std::vector<units::point> points_;
    std::mt19937_64 engine_;
};

int main()
{
    std::size_t const num_points = 60000;
    std::uint32_t const seed = 1234;
    int const num_steps = 50;

    std::ofstream output{"a.msg"};
    msgpack::packer<std::ostream> packer{&output};

    packer.pack_map(3);
    {
        packer.pack(std::string{"name"});
        packer.pack(std::string{"Lorem ipsum dolor sit amet"});

        packer.pack(std::string{"num_points"});
        packer.pack(std::uint32_t(num_points));

        packer.pack(std::string{"dimension"});
        packer.pack(std::uint32_t(units::point::dimension));
    }

    random_walk_simulator simulator{num_points, seed};

    packer.pack_array(std::uint32_t(num_steps));

    for (int step = 0; step < num_steps; ++step) {
        simulator.step_forward();

        packer.pack_bin(num_points * sizeof(float) * units::point::dimension);
        for (units::point const& point : simulator.points()) {
            float coords[units::point::dimension];
            for (unsigned axis = 0; axis < units::point::dimension; ++axis) {
                coords[axis] = static_cast<float>(point[axis].value());
            }
            // XXX: Raw binary serialization is bad!
            packer.pack_bin_body(reinterpret_cast<char const*>(coords), sizeof coords);
        }
    }
}
